sklearn/metrics/_classification.py

```diff
@@ -169,6 +172,69 @@ def _check_targets(y_true, y_pred, sample_weight=None):
     return y_type, y_true, y_pred, sample_weight
 
 
+def _one_hot_encoding_multiclass_target(y_true, labels, target_xp, target_device):
+    """Convert multi-class `y_true` into a one-hot encoded array and also ensure
+    that the encoded array is placed on the target API namespace and device.
+    Also return the classes provided by `LabelBinarizer` in additional to the
+    integer encoded array.
+    """
+    xp_y_true, is_y_true_array_api = get_namespace(y_true)
+
+    # For classification metrics both array API compatible and non array API
+    # compatible inputs are allowed for `y_true`. This is because arrays that
+    # store class labels as strings cannot be represented in namespaces other
+    # than Numpy. Thus to avoid unnecessary complexity, we always convert
+    # `y_true` to a Numpy array so that it can be processed appropriately by
+    # `LabelBinarizer` and then transfer the integer encoded output back to the
+    # target namespace and device.
+    if is_y_true_array_api:
+        y_true = _convert_to_numpy(y_true, xp=xp_y_true)
```

Comment:
Note that in a follow-up PR, we could change the label binarizer to avoid this conversion whenever the input is integer coded. This would involve adding array API support to `LabelBinarizer` when `sparse_output=False`.

But we can probably do that in a follow-up PR.

Note that this discussion caused https://github.com/scikit-learn/scikit-learn/pull/30439/files#r1875958580 to stall in the past but I think we can decouple the concerns.

Reply:
And it would be interesting to see the impact of not converting to numpy when running the benchmarks of https://github.com/scikit-learn/scikit-learn/pull/32422#issuecomment-3381082314 which uses integer class values in `y_true`.